### PR TITLE
add bytesToArray to Utils

### PR DESCRIPTION
**Pull request by dangell7**
## High Level Overview of Change

Add Bytes to Array Function to Utils for converting Hex String to bytes array

### Context of Change

Architecture was influenced by @keefertaylor 

### Type of Change

- [ X] New feature (non-breaking change which adds functionality)

## Before / After

Not Sure. Use whenever converting string hex to data

## Test Plan

All Tests Passed

## Future Tasks

Add to getTx

### CODE REVIEW

File: XpringKit/Utils.swift [hunk 1]
```diff
@@ -1,66 +1,132 @@
...
-  }
+    /// Convert the given transaction blob to a transaction hash.
+    ///
+    /// - Parameter transactionBlobHex: A hexadecimal encoded transaction blob.
+    /// - Returns: A hex encoded hash if the input was valid, otherwise undefined.
+    public static func toTransactionHash(transactionBlobHex: Hex) -> Hex? {
+        return javaScriptUtils.toTransactionHash(transactionBlobHex: transactionBlobHex)
+    }
+
+    /// Convert the given string to a bytes array UInt8.
```
Comment (keefertaylor) [inline XpringKit/Utils.swift]: "to a Uint8 byte array."

File: XpringKit/Utils.swift [hunk 2]
```diff
@@ -1,66 +1,132 @@
...
+    ///
+    /// - Parameter hex: A string.
+    /// - Returns: A byte array in [UInt8].
+    public static func toByteArray(hex: String ) -> [UInt8] {
```
Comment (keefertaylor) [inline XpringKit/Utils.swift]: There's a `Hex` type alias you can use instead of `String`

Comment (keefertaylor) [inline XpringKit/Utils.swift]: Nit: remove  space before `)`

File: XpringKit/Utils.swift [hunk 3]
```diff
@@ -1,66 +1,132 @@
...
+        // remove "-" from Hexadecimal
+        let hexString = hex.removeWord( "-" )
```
Comment (keefertaylor) [inline XpringKit/Utils.swift]: Remove spaces after `(` and before `)`. Here and elsewhere in this method. This is consistent with style throughout the existing code.

Comment (keefertaylor) [inline XpringKit/Utils.swift]: In practice, do we care about '-'. I don't think there's every any input into this functions that are dashes and this feels like an odd one-off check to me.

File: XpringKit/Utils.swift [hunk 4]
```diff
@@ -1,66 +1,132 @@
...
+        let size = hexString.count / 2
+        var result: [UInt8] = [UInt8]( repeating: 0, count: size ) // array with length = size
+        for i in stride( from: 0, to: hexString.count, by: 2 ) {
+            let subHexStr = hexString.subString( i, length: 2 )
+            guard let _uint8 = UInt8( subHexStr, radix: 16 ) else {
```
Comment (keefertaylor) [inline XpringKit/Utils.swift]: why prefix this with `_`

File: XpringKit/Utils.swift [hunk 5]
```diff
@@ -1,66 +1,132 @@
...
+                return []
+            }
+            result[ i / 2 ] = _uint8
+        }
+        return result
+    }
+}
+extension String {
+
+    func subString( _ from: Int, length: Int ) -> String {
```
Comment (keefertaylor) [inline XpringKit/Utils.swift]: I'd prefer we just use `prefix` to get the first two elements, rather than re-implementing a `substring` method:
https://developer.apple.com/documentation/swift/string/2894830-prefix

This method is essentially implementing [String's subscript](https://developer.apple.com/documentation/swift/string/2943910-subscript) method. Would prefer we use the official definitions provided by foundation rather than hand rolling our own.

File: XpringKit/Utils.swift [hunk 6]
```diff
@@ -1,66 +1,132 @@
...
+        var result = ""
+        for ( idx, char ) in self.enumerated() {
+            if idx >= from && idx < to {
+                result.append( char )
+            }
+        }
+        return result
+    }
+
+    func removeWord( _ word: String ) -> String {
```
Comment (keefertaylor) [inline XpringKit/Utils.swift]: It's a bit opaque what this method is doing. I think that you could just use [dropFirst](https://developer.apple.com/documentation/swift/string/2893343-dropfirst).

File: Tests/UtilsTest.swift [hunk 7]
```diff
@@ -3,242 +3,269 @@ import XpringKit
...
+        XCTAssertEqual(
+            hexBytesArray.toHex(),
+            "9A88C8548E03958FD97AF44AE5A8668896D195A70CF3FF3CB8E57096AA717135"
+        )
+    }
 
-    // THEN the hash is nil.
-    XCTAssertNil(transactionHash)
-  }
+    func testToBytesArrayInvalidTransactionHash() {
```
Comment (keefertaylor) [inline Tests/UtilsTest.swift]: This test feels odd to me. If we give an invalid input, I'd expect either:
- `nil` output
- an exception to be thrown for an invalid input 

I'd probably prefer to throw an exception rather than making the output optional but happy to hear your thoughts if you have a strong opinion.

I would expect 
```
> Utils.toByteArray(hex: "")
[]
```

There's no way to distinguish this case from the empty input case with the current implementation.

File: Tests/UtilsTest.swift [hunk 8]
```diff
@@ -3,242 +3,269 @@ import XpringKit
...
-    let transactionHash = Utils.toTransactionHash(transactionBlobHex: transactionBlobHex)
+    // MARK: - toBytesArray
 
-    // THEN the transaction blob is as expected.
-    XCTAssertEqual(
-      transactionHash,
-      "7B9F6E019C2A79857427B4EF968D77D683AC84F5A880830955D7BDF47F120667"
-    )
-  }
+    func testToBytesArrayValidTransactionHash() {
```
Comment (keefertaylor) [inline Tests/UtilsTest.swift]: Please add a test case for the empty string as well.

File: XpringKit/Utils.swift (hunk 5 above)
Comment (dangell7) [reply XpringKit/Utils.swift]: I couldn't get the prefix to work, It seems its a custom function. You're welcome to refactor the function.
